compiler/generator_php.c

```diff
@@ -0,0 +1,1406 @@
+    writef(g, ")~N", p);
+    write_block_start(g);
+    write_failure(g);
+    write_block_end(g);
+    g->unreachable = false;
+}
+
+/* if at limit fail */
+static void write_check_limit(struct generator * g, struct node * p) {
+    if (p->mode == m_forward) {
+        write_failure_if(g, "$this->cursor >= $this->limit", p);
+    } else {
+        write_failure_if(g, "$this->cursor <= $this->limit_backward", p);
+    }
+}
+
+/* Formatted write. */
+static void writef(struct generator * g, const char * input, struct node * p) {
+    (void)p;
+    int i = 0;
+
+    while (input[i]) {
+        int ch = input[i++];
+        if (ch != '~') {
+            write_char(g, ch);
+            continue;
+        }
+        ch = input[i++];
+        switch (ch) {
+            case '~': write_char(g, '~'); continue;
+            case 'f': write_block_start(g);
+                      write_failure(g);
+                      g->unreachable = false;
+                      write_block_end(g);
+                      continue;
+            case 'M': write_margin(g); continue;
+            case 'N': write_newline(g); continue;
+            case '{': write_block_start(g); continue;
+            case '}': write_block_end(g); continue;
+            case 'S': {
+                int j = input[i++] - '0';
+                if (j < 0 || j > (int)(sizeof(g->S) / sizeof(g->S[0]))) {
+                    printf("Invalid escape sequence ~%c%c in writef(g, \"%s\", p)\n",
+                           ch, input[i - 1], input);
+                    exit(1);
+                }
+                write_string(g, g->S[j]);
+                continue;
+            }
+            case 'B': {
+                int j = input[i++] - '0';
+                if (j < 0 || j > (int)(sizeof(g->B) / sizeof(g->B[0])))
+                    goto invalid_escape2;
+                write_s(g, g->B[j]);
+                continue;
+            }
+            case 'I': {
+                int j = input[i++] - '0';
+                if (j < 0 || j > (int)(sizeof(g->I) / sizeof(g->I[0])))
+                    goto invalid_escape2;
+                write_int(g, g->I[j]);
+                continue;
+            }
+            case 'V':
+                write_varref(g, p->name);
+                continue;
+            case 'W':
+                write_varname(g, p->name);
+                continue;
+            case 'L':
+                write_literal_string(g, p->literalstring);
+                continue;
+            case '+': g->margin++; continue;
+            case '-': g->margin--; continue;
+            case 'n': write_string(g, g->options->name); continue;
+            case 'P': write_string(g, g->options->parent_class_name); continue;
+            case 'C': { // Constant.
+                w(g, "/* TODO */ const"); // ??
+                continue;
+            }
+            case 'D': { // Declare variable.
+                w(g, "/* TODO ?*/" ); // ??
+                continue;
+            }
```

Comment:
`~C` and `~D` have now been scrapped (they were JS specific, but we now only support generating ES6 so can just use `const` and `let` unconditionally).